### PR TITLE
Add a delete confirm to the user list

### PR DESCRIPTION
**Pull request by huwshimi**
Done:
- Show a delete confirmation when deleting user from the user list. Fixes: https://github.com/canonical-web-and-design/maas-squad/issues/1354. Fixes: https://github.com/canonical-web-and-design/maas-ui/issues/53.

QA:
- View the list of users.
- Click the delete icon next to a user.
- A confirmation should appear.
- Click cancel, it should disappear.

### CODE REVIEW

File: ui/src/app/settings/containers/Users/Users.scss [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  padding-left: 0.5rem;
+  padding-right: 0.5rem;
+}
+
+.user-list .p-table-expanding__panel {
+  padding-top: 0;
+}
+
+.p-table-expanding th.user-list__cell--small,
+.p-table-expanding td.user-list__cell--small {
```
Comment (anthonydillon) [inline ui/src/app/settings/containers/Users/Users.scss]: Could this modifier be added to the table-expanding pattern scss. If someone needs to modify this styles the first place they will look will be the pattern file.

File: ui/src/app/settings/containers/Users/Users.scss [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  flex: 0 0 6rem;
+}
+
+.user-list__expanded-row {
+  background-color: #fff;
```
Comment (anthonydillon) [inline ui/src/app/settings/containers/Users/Users.scss]: This is also something we will want across all tables. Could you add to the table pattern locally here. For example: `p-table__row.is-active` then we could propose this to Vanilla.

File: ui/src/app/settings/containers/Users/Users.js [hunk 3]
```diff
@@ -14,54 +15,85 @@ import Pagination from "app/base/components/Pagination";
...
+          content: <Link to={`/users/${user.id}`}>{user.username}</Link>,
+          role: "rowheader"
+        },
+        { content: user.email },
+        { content: 37, className: "u-align--right user-list__cell--small" },
+        { content: "Local", className: "user-list__cell--small" },
+        { content: "12 mins ago" },
+        {
+          content: user.is_superuser ? "Admin" : null,
+          className: "user-list__cell--small"
```
Comment (anthonydillon) [inline ui/src/app/settings/containers/Users/Users.js]: In my experience design want more control over column widths which lead use to setting custom widths to all table columns. Which you can see here: https://github.com/maas/maas/blob/master/src/maasserver/static/scss/_tables.scss

File: ui/src/app/settings/containers/Users/Users.js [hunk 4]
```diff
@@ -14,54 +15,85 @@ import Pagination from "app/base/components/Pagination";
...
+        },
+        {
+          content: user.sshkeys_count,
+          className: "u-align--right user-list__cell--small"
+        },
+        {
+          content: (
+            <>
```
Comment (anthonydillon) [inline ui/src/app/settings/containers/Users/Users.js]: Is this meant to have an empty element?

File: ui/src/app/settings/containers/Users/Users.js [hunk 5]
```diff
@@ -14,54 +15,85 @@ import Pagination from "app/base/components/Pagination";
...
+          className: "u-align--right user-list__cell--small"
+        }
+      ],
+      expanded: expanded,
+      expandedContent: expanded && (
+        <div className="user-list__expanded-content">
+          <Row>
+            <Col size="8">
+              Are you sure you want to delete user "{user.username}"?{" "}
+              <span className="u-text--light">
```
Comment (anthonydillon) [inline ui/src/app/settings/containers/Users/Users.js]: This seems like the main message. I don't think we should be making it subtle.

File: ui/src/app/settings/containers/Users/Users.scss (hunk 2 above)
Comment (anthonydillon) [reply ui/src/app/settings/containers/Users/Users.scss]: Could you use Vanilla colour variables please. `$color-x-light`

File: ui/src/app/settings/containers/Users/Users.js [hunk 6]
```diff
@@ -14,54 +15,85 @@ import Pagination from "app/base/components/Pagination";
...
-          </>
-        ),
-        className: "u-align--right"
+                <i className="p-icon--edit">Edit</i>
+              </Button>
+              {user.id !== authUser.id && (
+                <Button
+                  appearance="base"
+                  className="is-small user-list__button"
+                  onClick={() => setExpandedId(user.id)}
```
Comment (anthonydillon) [inline ui/src/app/settings/containers/Users/Users.js]: Your giving each row a lot of power here. They shouldn't have the ability to set other row. Maybe think about moving this state to the table component itself? And then just passing a click function to the rows?

File: ui/src/app/settings/containers/Users/Users.js [hunk 7]
```diff
@@ -3,6 +3,7 @@ import { Link } from "react-router-dom";
 import PropTypes from "prop-types";
 import React, { useState } from "react";
 
+import "./Users.scss";
```
[the remark this answers is not in the document]
Comment (huwshimi) [reply ui/src/app/settings/containers/Users/Users.js]: This is a great way of managing css specific to a component (https://facebook.github.io/create-react-app/docs/adding-a-stylesheet) and within the SCSS file you can import the vanilla bits you need `@import "~vanilla-framework/scss/settings_colors";`.

In this case, you're right in that these styles should not be specific to the user list.

File: ui/src/app/settings/containers/Users/Users.scss [hunk 8]
```diff
@@ -0,0 +1,29 @@
+@import "src/scss/vanilla";
+
+.user-list__button.p-button--base {
```
[the remark this answers is not in the document]
Comment (huwshimi) [reply ui/src/app/settings/containers/Users/Users.scss]: Thanks, I've made this a global rule and filed: https://github.com/canonical-web-and-design/vanilla-framework/issues/2463

File: ui/src/app/settings/containers/Users/Users.scss [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+  // Stop the icons from adding extra height to the row:
+  margin-top: -0.25rem;
+  margin-bottom: -0.25rem;
+  padding-left: 0.5rem;
+  padding-right: 0.5rem;
+}
+
+.user-list .p-table-expanding__panel {
+  padding-top: 0;
```
[the remark this answers is not in the document]
Comment (huwshimi) [reply ui/src/app/settings/containers/Users/Users.scss]: You're right, this should be for all the tables on a dark background.

File: ui/src/app/settings/containers/Users/Users.scss (hunk 1 above)
Comment (huwshimi) [reply ui/src/app/settings/containers/Users/Users.scss]: Really I think we need more control over the column widths than this from looking at the design, do you have any suggestions about a good way to set the width of each column?

Comment (huwshimi) [reply ui/src/app/settings/containers/Users/Users.scss]: Oops, you answered below.

File: ui/src/app/settings/containers/Users/Users.js (hunk 4 above)
Comment (huwshimi) [reply ui/src/app/settings/containers/Users/Users.js]: This is shorthand for `React.Fragment` which can be used when you need to wrap multiple elements but don't need anything to appear in the DOM.

File: ui/src/app/settings/containers/Users/Users.js (hunk 5 above)
Comment (huwshimi) [reply ui/src/app/settings/containers/Users/Users.js]: This matches the design here: https://app.zeplin.io/project/5ce5423a25863c1e0d7abb27/screen/5d36e2233615fbb2789ce87f.

Sorry just realised I didn't link to the design in the PR.

File: ui/src/app/settings/containers/Users/Users.js (hunk 6 above)
Comment (huwshimi) [reply ui/src/app/settings/containers/Users/Users.js]: There are a couple of issues there, one is that we don't have a consistent element that opens the row (in this case it is the trash icon).

The second is that we need to be able to control that state outside the table. For this list when the user confirms the delete then we need to close that row.

File: ui/src/app/settings/containers/Users/Users.scss (hunk 2 above)
Comment (huwshimi) [reply ui/src/app/settings/containers/Users/Users.scss]: Done.

File: ui/src/app/settings/containers/Users/Users.js (hunk 3 above)
Comment (huwshimi) [reply ui/src/app/settings/containers/Users/Users.js]: Done, thanks for that, I've updated these to match the design.

File: ui/src/app/base/components/Loader/Loader.scss [hunk 10]
```diff
@@ -1,4 +1,4 @@
-@import "src/scss/vanilla";
+@import "~vanilla-framework/scss/settings_spacing";
```
Comment (huwshimi) [inline ui/src/app/base/components/Loader/Loader.scss]: These are all drive-by fixes to only include the vanilla bits we need.

**Comment by squidsoup** [comment]
QA +1, thanks @huwshimi!
